In [20]:
import pandas as pd
import numpy as np
import nltk
import string
from nltk.tokenize import word_tokenize
from nltk.corpus import stopwords

In [15]:
yt_spam = pd.read_csv("Youtube04-Eminem.csv")
yt_spam.head()

COMMENT_ID         AUTHOR  \
0    z12rwfnyyrbsefonb232i5ehdxzkjzjs2    Lisa Wellas   
1  z130wpnwwnyuetxcn23xf5k5ynmkdpjrj04   jason graham   
2    z13vsfqirtavjvu0t22ezrgzyorwxhpf3     Ajkal Khan   
3  z12wjzc4eprnvja4304cgbbizuved35wxcs  Dakota Taylor   
4    z13xjfr42z3uxdz2223gx5rrzs3dt5hna    Jihad Naser   

                         DATE  \
0                         NaN   
1  2015-05-29T02:26:10.652000   
2                         NaN   
3  2015-05-29T02:13:07.810000   
4                         NaN   

                                             CONTENT  CLASS  
0          +447935454150 lovely girl talk to me xxx      1  
1    I always end up coming back to this song<br />      0  
2  my sister just received over 6,500 new <a rel=...      1  
3                                              Cool      0  
4                     Hello I&#39;am from Palastine      1

In [16]:
yt_spam.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 448 entries, 0 to 447
Data columns (total 5 columns):
 #   Column      Non-Null Count  Dtype 
---  ------      --------------  ----- 
 0   COMMENT_ID  448 non-null    object
 1   AUTHOR      448 non-null    object
 2   DATE        203 non-null    object
 3   CONTENT     448 non-null    object
 4   CLASS       448 non-null    int64 
dtypes: int64(1), object(4)
memory usage: 17.6+ KB


In [17]:
nltk.download('punkt')

[nltk_data] Downloading package punkt to
[nltk_data]     C:\Users\preay\AppData\Roaming\nltk_data...
[nltk_data]   Package punkt is already up-to-date!


True

In [18]:
yt_spam.CONTENT

0              +447935454150 lovely girl talk to me xxx
1        I always end up coming back to this song<br />
2      my sister just received over 6,500 new <a rel=...
3                                                  Cool
4                         Hello I&#39;am from Palastine
                             ...                        
443             SUBSCRIBE TO MY CHANNEL X PLEASE!. SPARE
444    Check out my videos guy! :) Hope you guys had ...
445    3 yrs ago I had a health scare but thankfully ...
446         Rihanna looks so beautiful with red hair ;)
447    857.482.940 views AWESOME !!!!!!!!!!!!!!!!!!!!...
Name: CONTENT, Length: 448, dtype: object

In [24]:
nltk.download('stopwords')

[nltk_data] Downloading package stopwords to
[nltk_data]     C:\Users\preay\AppData\Roaming\nltk_data...
[nltk_data]   Unzipping corpora\stopwords.zip.


True

In [25]:
def processing_test(text):
    
    
    nopunc= [char for char in text if char not in string.punctuation]
    nopunc= ''.join(nopunc)
    
    clean_words =[word for word in nopunc.split() if word.lower() not in stopwords.words('english')]
    
    return clean_words

In [26]:
yt_spam.CONTENT.apply(processing_test
                     )

0               [447935454150, lovely, girl, talk, xxx]
1                 [always, end, coming, back, songbr, ]
2      [sister, received, 6500, new, relnofollow, cla...
3                                                [Cool]
4                             [Hello, I39am, Palastine]
                             ...                        
443               [SUBSCRIBE, CHANNEL, X, PLEASE, SPARE]
444        [Check, videos, guy, Hope, guys, good, laugh]
445    [3, yrs, ago, health, scare, thankfully, I’m, ...
446            [Rihanna, looks, beautiful, red, hair, ]
447                       [857482940, views, AWESOME, ]
Name: CONTENT, Length: 448, dtype: object

In [27]:
from sklearn.feature_extraction.text import CountVectorizer

In [47]:
message= CountVectorizer(analyzer=processing_test).fit_transform(yt_spam.CONTENT)
message.shape

(448, 1980)

In [33]:
from sklearn.model_selection import train_test_split
x_train,x_test,y_train,y_test =train_test_split(message,yt_spam.CLASS,test_size=0.2)

In [50]:
from sklearn.naive_bayes import MultinomialNB
classifier=MultinomialNB().fit(x_train,y_train)


In [51]:
y_preds=classifier.predict(x_test)

In [59]:
from sklearn.metrics import classification_report,confusion_matrix,accuracy_score

print(classification_report(y_test,y_preds))
print(confusion_matrix(y_test,y_preds))
print()
print(accuracy_score(y_test,y_preds))

              precision    recall  f1-score   support

           0       0.83      0.98      0.90        49
           1       0.97      0.76      0.85        41

    accuracy                           0.88        90
   macro avg       0.90      0.87      0.87        90
weighted avg       0.89      0.88      0.88        90

[[48  1]
 [10 31]]

0.8777777777777778


# so we got an accuracy of 87% to predict weather a comment on yt video is spam or not